fastcdc/fastcdc.h

```diff
@@ -239,30 +160,42 @@ class ChunkerTmpl {
       len = cfg_.max_size;
     }
 
+    // Initialize the regression length to len (the end) and the regression
+    // mask to an empty bitmask (match any hash).
+    size_t rc_len = len;
+    T rc_mask = 0;
+
     // Init hash to all 1's to avoid zero-length chunks with min_size=0.
-    uint64_t hash = UINT64_MAX;
-    // Skip the first min_size bytes, but "warm up" the rolling hash for 64
-    // rounds to make sure the 64-bit hash has gathered full "content history".
-    size_t i = cfg_.min_size > 64 ? cfg_.min_size - 64 : 0;
+    T hash = (T)-1;
+    // Skip the first min_size bytes, but "warm up" the rolling hash for enough
+    // rounds to make sure the hash has gathered full "content history".
+    size_t i = cfg_.min_size > khashbits_ ? cfg_.min_size - khashbits_ : 0;
     for (/*empty*/; i < cfg_.min_size; ++i) {
       hash = (hash << 1) + gear[data[i]];
     }
-    for (uint32_t stg = 0; stg < mask_stages && i < len; stg++) {
-      uint64_t mask = stages_[stg].mask;
-      size_t barrier = std::min(len, stages_[stg].barrier);
-      for (/*empty*/; i < barrier; ++i) {
-        if (!(hash & mask)) {
+    for (/*empty*/; i < len; ++i) {
+      if (!(hash & rc_mask)) {
+        if (hash <= kthreshold_) {
+          // This hash matches the target length hash criteria, return it.
           return i;
+        } else {
+          // This is a better regression point. Set it as the new rc_len and
+          // update rc_mask to check as many MSBits as this hash would pass.
+          rc_len = i;
+          for (rc_mask = (T)-1; hash & rc_mask; rc_mask <<= 1)
+            ;
         }
-        hash = (hash << 1) + gear[data[i]];
       }
+      hash = (hash << 1) + gear[data[i]];
     }
-    return i;
+    // Return best regression point we found or the end if it's better.
+    return (hash & rc_mask) ? rc_len : i;
   }
 
+  static constexpr size_t khashbits_ = sizeof(T) * 8;
   const Config cfg_;
   const ChunkFoundHandler handler_;
-  MaskStage stages_[mask_stages];
+  T kthreshold_;
```

Comment:
Nit: We generally use the k notation only if the variable is actually a const. Maybe just threshold_ or make it const T (which you can do if you compute it after the : in the constructor, though that's a bit ugly).